File: android/app/src/main/java/com/voicecambuffer/circularbuffer/FrameRingBuffer.kt
package com.voicecambuffer.circularbuffer

import java.util.ArrayDeque

/**
 * Pure, Android-framework-free ring buffer of already-encoded frames.
 *
 * Frames are expected to arrive in non-decreasing presentation-time order
 * (as MediaCodec emits them). The buffer evicts frames older than
 * [windowUs] every time a new frame is appended, so memory stays bounded
 * regardless of how long buffering runs.
 *
 * This class intentionally has zero Android SDK dependency so it can run
 * under plain JVM JUnit tests (no instrumentation/emulator needed).
 */
class FrameRingBuffer(windowUs: Long) {

    var windowUs: Long = windowUs
        private set

    private val frames = ArrayDeque<EncodedFrame>()

    val size: Int
        get() = frames.size

    fun updateWindow(newWindowUs: Long) {
        windowUs = newWindowUs
        evictOldFrames()
    }

    @Synchronized
    fun add(frame: EncodedFrame) {
        frames.addLast(frame)
        evictOldFrames()
    }

    private fun evictOldFrames() {
        if (frames.isEmpty()) return
        val newestPts = frames.last.presentationTimeUs
        while (frames.isNotEmpty() && newestPts - frames.first.presentationTimeUs > windowUs) {
            frames.removeFirst()
        }
    }

    @Synchronized
    fun clear() {
        frames.clear()
    }

    /**
     * Returns an immutable snapshot of all frames currently held, starting
     * at the oldest available keyframe (so the result can be muxed into a
     * decodable clip). If the buffer holds no keyframe at all (e.g. GOP
     * length longer than the buffer window), every currently buffered frame
     * is returned instead of failing -- callers must not crash or throw in
     * this case, they simply get a clip that starts mid-GOP.
     */
    @Synchronized
    fun snapshotFromOldestKeyframe(): List<EncodedFrame> {
        val all = frames.toList()
        val firstKeyframeIndex = all.indexOfFirst { it.isKeyFrame }
        return if (firstKeyframeIndex <= 0) all else all.subList(firstKeyframeIndex, all.size)
    }

    @Synchronized
    fun snapshot(): List<EncodedFrame> = frames.toList()
}

File: ios/CircularBuffer/FrameRingBuffer.swift
import Foundation

/// Pure, AVFoundation-free ring buffer of already-encoded frames.
///
/// Frames are expected to arrive in non-decreasing presentation-time order.
/// The buffer evicts frames older than `windowUs` every time a new frame is
/// appended, so memory stays bounded regardless of how long buffering runs.
///
/// This type has zero AVFoundation/UIKit dependency so it can run under
/// plain XCTest unit tests (no simulator camera or device needed).
final class FrameRingBuffer {
    private(set) var windowUs: Int64
    private var frames: [EncodedFrame] = []
    private let lock = NSLock()

    init(windowUs: Int64) {
        self.windowUs = windowUs
    }

    var size: Int {
        lock.lock(); defer { lock.unlock() }
        return frames.count
    }

    func updateWindow(_ newWindowUs: Int64) {
        lock.lock(); defer { lock.unlock() }
        windowUs = newWindowUs
        evictOldFrames()
    }

    func add(_ frame: EncodedFrame) {
        lock.lock(); defer { lock.unlock() }
        frames.append(frame)
        evictOldFrames()
    }

    /// Must be called while holding `lock`.
    private func evictOldFrames() {
        guard let newestPts = frames.last?.presentationTimeUs else { return }
        while let oldestPts = frames.first?.presentationTimeUs, newestPts - oldestPts > windowUs {
            frames.removeFirst()
        }
    }

    func clear() {
        lock.lock(); defer { lock.unlock() }
        frames.removeAll()
    }

    func snapshot() -> [EncodedFrame] {
        lock.lock(); defer { lock.unlock() }
        return frames
    }

    /// Returns every buffered frame starting at the oldest keyframe, so the
    /// result can be fed to AVAssetWriter as a decodable clip. If the buffer
    /// holds no keyframe at all (GOP longer than the buffer window) every
    /// currently buffered frame is returned instead of failing -- callers
    /// must never crash or throw here, they just get a clip that starts
    /// mid-GOP.
    func snapshotFromOldestKeyframe() -> [EncodedFrame] {
        lock.lock(); defer { lock.unlock() }
        guard let firstKeyframeIndex = frames.firstIndex(where: { $0.isKeyFrame }), firstKeyframeIndex > 0 else {
            return frames
        }
        return Array(frames[firstKeyframeIndex...])
    }
}
